### ISSUE TITLE
Handling 32-bit overflow in jumptables

### ISSUE THREAD

**Issue by aaaaaa123456789** [fstirlitz/c-bsp#1]
It seems like I forgot to specify the behavior of snippets like these:

```
  set #0, -2
  jump .jt_inst

  dw .exit ;-8
  nop ;-4
  nop ;-3
.jt_inst
  jumptable #0 ;-2

.exit
  exit 0
```

```
  set #0, 0x40000000
  jumptable #0
  dw .exit

.exit
  exit 0
```

In my implementation, they both fail due to overflowing the program counter; by reading yours, it would seem like the overflow is just ignored and they both exit with zero.

I haven't been able to test your implementation yet, so I have to ask:

1. What's the current behavior?
2. What's the _intended_ behavior?
3. If they do exit 0, shouldn't they fail instead?

**Comment by aaaaaa123456789** [comment]
(also, since I see no way of contacting you, would you mind emailing me? My email is in my profile. There are a few things I'd like to comment on, but they are more conversational in nature and I doubt they belong in an issue)

**Comment by fstirlitz** [comment]
The overflow is indeed ignored.

```
$ ./runtest -v tests/jumptable_wraparound.test 
=== Compiling ===
=== Disassembling ===
=== Disassembly ===
                        set                #0, 0xfffffffe
label_00000006:
                        jump               label_00000011
                        db                 0x13
                        db                 0x00
                        db                 0x00
                        db                 0x00
                        db                 0x00
                        db                 0x00
label_00000011:
                        jumptable          #0
                        dw                 label_00000006
                        db                 0x00

=== Compiling again ===
=== Verifying ===
=== Running ===
=== Standard error ===
[  0/000001] 00000000  8400feffffff            set               #0, 0xfffffffe
                                               #0   = 0xfffffffe
                                               ------
[  0/000002] 00000006  0211000000              jump              0x00000011
                                               ------
                                               #0   = 0xfffffffe
[  0/000003] 00000011  8300                    jumptable         #0
                                               ------
[  0/000004] 00000013  0600000000              exit              0x00000000
./bsp: /tmp/bsptest-9ppygheh/bsp0[+0x13]: patching successful, exit code 0
./bsp: /tmp/bsptest-9ppygheh/bsp0: 4 instructions executed
./bsp: /tmp/bsptest-9ppygheh/bsp0: 0 backward jumps

=== Expected output ===

=== Expected final data ===

=== Actual final data matches ===
=== Expected a fatal error ===
=== Actual exit code: 0 ===
```

```
$ ./runtest -v tests/jumptable_wraparound_zero.test 
=== Compiling ===
=== Disassembling ===
=== Disassembly ===
                        set                #0, 0x40000000
                        jumptable          #0
                        dw                 label_0000000c
label_0000000c:
                        exit               0x00000000

=== Compiling again ===
=== Verifying ===
=== Running ===
=== Standard error ===
[  0/000001] 00000000  840000000040            set               #0, 0x40000000
                                               #0   = 0x40000000
                                               ------
                                               #0   = 0x40000000
[  0/000002] 00000006  8300                    jumptable         #0
                                               ------
[  0/000003] 0000000c  0600000000              exit              0x00000000
./bsp: /tmp/bsptest-trco_6uz/bsp0[+0xc]: patching successful, exit code 0
./bsp: /tmp/bsptest-trco_6uz/bsp0: 3 instructions executed
./bsp: /tmp/bsptest-trco_6uz/bsp0: 0 backward jumps

=== Expected output ===

=== Expected final data ===

=== Actual final data matches ===
=== Expected a fatal error ===
=== Actual exit code: 0 ===
```

I agree that this should be considered an oversight, not intended behaviour. Good catch. (And here I thought my test suite was covering everything.) While I realise die-hard assembly programmers may sometimes find such quirks useful, I think overflowing offsets are more likely to be indicative of a logic error. The 'negative index' case seems too cumbersome to invoke to be useful, and if you want to use the upper two bits of a jump table index for flag bits or something, you can always mask them out before executing `jumptable`. It's not like patch scripts are performance-constrained. (It's not likely you'll run out of registers when keeping flags separate either.)

> also, since I see no way of contacting you

The address in my commits should be functional; feel free to use that. However, you probably shouldn't expect a quick reply, I am quite busy as of late.

**Comment by fstirlitz** [comment]
Fixed in ee877c955d7144a6f2da6001b38db882e4425f67.
